### PR TITLE
[FX] Make Graphs immutable and make GraphModule recompile after assigning graph

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44830 [FX] Make Graphs immutable and make GraphModule recompile after assigning graph**

Differential Revision: [D23743850](https://our.internmc.facebook.com/intern/diff/D23743850)

### CODE REVIEW

**Review by zdevito — COMMENTED** [review]
I don't think we should expose generate forward. It makes more sense that assignment to the `.graph` property automatically regenerates forward. `graph` itself should be immutable. This patch should also enforce the immutability of `graph` after it has had an output declared.

**Comment by jamesr66a** [comment]
@zdevito  how do we enforce immutability of `Graph`? Subclass `list` and disable all methods except append and getitem or something?

**Comment by jamesr66a** [comment]
@zdevito updated. I don't particularly like how this breaks editing after copy/deepcopy

File: torch/fx/graph.py [hunk 1]
```diff
@@ -64,10 +65,46 @@ def map_arg(a: Argument, fn: Callable[[Node], Argument]) -> Argument:
...
 class Graph:
-    def __init__(self):
-        self.nodes : List[Node] = []
+    def __init__(self, from_graph : Optional['Graph'] = None):
         self._used_names : Dict[str, int] = {}  # base name -> number
+        if from_graph:
+            self.nodes : FinalizableList = FinalizableList()
+            val_map = {}
+            for node in from_graph.nodes:
+                val_map[node] = self.node_copy(node, lambda n: val_map[n])
```
Comment (dzhulgakov) [inline torch/fx/graph.py:105]: btw, should `node_copy` be delegate-customizable too? I'm thinking of cases when extra metadata is hanging off the node and we want to copy it

File: torch/fx/graph.py [hunk 2]
```diff
@@ -130,6 +167,7 @@ def node_copy(self, node: Node, arg_transform: Callable[[Node], Argument] = lamb
     def output(self, result: Argument):
         self.result = result
         self._mark_uses(result)
+        self.nodes.finalize()
```
Comment (dzhulgakov) [inline torch/fx/graph.py:170]: btw, have you considered just doing `self.nodes = tuple(self.nodes)`? :) What is the benefit of FinalizableList over Tuple?

File: torch/fx/graph.py (hunk 1 above)
Comment (jamesr66a) [reply torch/fx/graph.py:105]: Delegate is now Tracer and only affects how tracing occurs

File: torch/fx/graph.py (hunk 2 above)
Comment (jamesr66a) [reply torch/fx/graph.py:170]: Every time you append to the tuple you'd have to create a new data structure? Also that doesn't match with the specification above, of immutable after an output has been specified

File: torch/fx/graph.py [hunk 3]
```diff
@@ -64,10 +65,46 @@ def map_arg(a: Argument, fn: Callable[[Node], Argument]) -> Argument:
     else:
         return a
 
+class FinalizableList(list):
```
Comment (zdevito) [inline torch/fx/graph.py:68]: Not sure all this infrastructure is worth it. We can just change `nodes` to be a property and then editing nodes doesn't change the graph.

File: torch/fx/graph.py [hunk 4]
```diff
@@ -64,10 +65,46 @@ def map_arg(a: Argument, fn: Callable[[Node], Argument]) -> Argument:
...
+            if self.finalized:
+                self._raise_finalized()
+            getattr(super(FinalizableList, self), method)(*args, **kwargs)
+        return wrapper
+    setattr(FinalizableList, method, scope(method))
+
 class Graph:
-    def __init__(self):
-        self.nodes : List[Node] = []
+    def __init__(self, from_graph : Optional['Graph'] = None):
```
Comment (zdevito) [inline torch/fx/graph.py:99]: It's a big odd to put this in the constructor. A graph could have a function `copy_graph` similar to copy node that would copy/append the entire graph. Then someone wanting to append to a graph can:

```
b = Graph()
a.copy_graph(b, input_fn) # same structure as copy node but copies an entire graph
```

Appending to a graph only is a specific use case, the general one is editing a graph (i.e. doing node-by-node transforms) and then splices graphs into eachother (graph_copy).

File: torch/fx/graph.py (hunk 1 above)
Comment (dzhulgakov) [reply torch/fx/graph.py]: Then what is the recommended mechanism fo attaching more stuff to nodes in the graph and preserving them? Should we copy entries in the `__dict__` (though it's a bit hacky)

File: torch/fx/graph.py (hunk 2 above)
Comment (dzhulgakov) [reply torch/fx/graph.py]: I mean to change type from `list` to `tuple` only when you want to finalize. Afaiu, list and tuple are mostly indistinguishable from the consumption standpoint.

File: torch/fx/graph.py (hunk 1 above)
Comment (jamesr66a) [reply torch/fx/graph.py]: Maybe we could add `extra_info : Dict[str, Any]` to Node that gets copied over?

**Review by zdevito — APPROVED** [review]

**Pull request closed without merging**